### PR TITLE
[SPARK-35678][ML] add a common softmax function

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
add softmax function in utils

### Why are the changes needed?
it can be used in multi places

### Does this PR introduce _any_ user-facing change?
NO

### How was this patch tested?
existing testsuites

### CODE REVIEW

**Comment by zhengruifeng** [comment]
ping @srowen

File: mllib-local/src/main/scala/org/apache/spark/ml/impl/Utils.scala [hunk 1]
```diff
@@ -94,4 +95,43 @@ private[spark] object Utils {
...
+        return
+      } else if (value > maxValue) {
+        maxValue = value
+      }
+      i += 1
+    }
+
+    var sum = 0.0
+    i = 0
+    if (maxValue > 0) {
```
Comment (srowen) [inline mllib-local/src/main/scala/org/apache/spark/ml/impl/Utils.scala:119]: Hm, would this not be valuable even if the max were very small? avoids underflow in exp in the same way it avoids overflow when the max is large. This avoids two branches

Comment (zhengruifeng) [reply mllib-local/src/main/scala/org/apache/spark/ml/impl/Utils.scala:119]: sounds reasonable

Comment (zhengruifeng) [reply mllib-local/src/main/scala/org/apache/spark/ml/impl/Utils.scala:119]:
```py
def softmax(X, copy=True):
    """
    Calculate the softmax function.

    The softmax function is calculated by
    np.exp(X) / np.sum(np.exp(X), axis=1)

    This will cause overflow when large values are exponentiated.
    Hence the largest value in each row is subtracted from each data
    point to prevent this.

    Parameters
    ----------
    X : array-like of float of shape (M, N)
        Argument to the logistic function.

    copy : bool, default=True
        Copy X or not.

    Returns
    -------
    out : ndarray of shape (M, N)
        Softmax function evaluated at every point in x.
    """
    if copy:
        X = np.copy(X)
    max_prob = np.max(X, axis=1).reshape((-1, 1))
    X -= max_prob
    np.exp(X, X)
    sum_prob = np.sum(X, axis=1).reshape((-1, 1))
    X /= sum_prob
    return X
```

softmax in scikit-learn does not check whether the maxvalue is positive or not

**Review by srowen — COMMENTED** [review]
Looking good. I actually wonder if it might change some test results - probably for the better if anything

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
You can merge when ready

**Comment by zhengruifeng** [comment]
Merged to master, thanks @srowen for reviewing!
